[doc] Update documentaion on meta programming for #1374 (#1389)
taichi-dev/taichi: docs/meta.rst

    @ti.kernel
    def copy(x: ti.template(), y: ti.template()):
        for i in x:
            y[i] = x[i]

    a = ti.var(ti.f32, 4)
    b = ti.var(ti.f32, 4)
    c = ti.var(ti.f32, 12)
    d = ti.var(ti.f32, 12)
    copy(a, b)
    copy(c, d)

As shown by the above example, template programming may enable us to reuse our
code and improve better flexibility.

Dimensionality-independent programming using grouped indices
------------------------------------------------------------

However, the ``copy`` template shown above is not perfect, i.e., it can only be
used to copy 1D tensors. What if we want to copy 2D tensors? Do we have to write
another kernel?

.. code-block:: python

    @ti.kernel
    def copy2d(x: ti.template(), y: ti.template()):
        for i, j in x:
            y[i, j] = x[i, j]

Not necessary! Taichi provides ``ti.grouped`` syntax which enable you to get
loop indices into a grouped vector, therefore unify different dimensionalities.
For example:

.. code-block:: python

    @ti.kernel
    def copy(x: ti.template(), y: ti.template()):
        for I in ti.grouped(y):
            # I is a vector with same dimensionality with x and data type i32

>             # If y is 0D, then I = None

Author: Oh! Didn't know that:
```suggestion
            # If y is 0D, then I = ti.Vector([])
```
But actually we may preserve this so that people who know `x[None]` can quickly got our idea?

>             # If y is 0D, then I = None

Reviewer: Good point. Maybe we say `then I = ti.Vector([]), which is equivalent to "None" when used in x[I]`?

>             # If y is 0D, then I = None

Author: ```suggestion
            # If y is 0D, then I = ti.Vector([]), which is equivalent to `None` when used in x[I]
```

            # If y is 1D, then I = ti.Vector([i])
            # If y is 2D, then I = ti.Vector([i, j])
            # If y is 3D, then I = ti.Vector([i, j, k])
            # ...
            x[I] = y[I]

    @ti.kernel
    def array_op(x: ti.template(), y: ti.template()):
        # if tensor x is 2D:

>         for I in ti.grouped(x): # I is simply a 2D vector with x data type i32

Author: What does `a 2D vector with x` mean? I don't understand.. wasn't the data type always i32 since it's an index?

>         for I in ti.grouped(x): # I is simply a 2D vector with x data type i32

Reviewer: I don't understand `x` either so I removed it...

>         for I in ti.grouped(x): # I is simply a 2D vector with x data type i32

Author: Oh sorry! I didn't distinguish red and green..

            y[I + ti.Vector([0, 1])] = I[0] + I[1]

        # then it is equivalent to:
        for i, j in x:
            y[i, j + 1] = i + j

Tensor meta data
----------------

Sometimes it will be useful to get the data type (``tensor.dtype``) and shape (``tensor.shape``) of tensors.
These attributes can be accessed in both Taichi kernels and python scripts.

.. code-block:: python

  @ti.func
  def print_tensor_info(x: ti.template()):
    print('Tensor dimensionality is', len(x.shape))
    for i in ti.static(range(len(x.shape))):
      print('Size alone dimension', i, 'is', x.shape[i])
    ti.static_print('Tensor data type is', x.dtype)

See :ref:`scalar_tensor` for more details.

.. note::

    For sparse tensors, the full domain shape will be returned.

Matrix & vector meta data
-------------------------

> Sometimes it will also be useful to get the matrix column and row numbers when
> you want to write dimensionality-independent code, such as reusing code between
> 2D/3D physical simulations.
> 
> ``matrix.m`` equals to the column number of matrix, while ``matrix.n`` equals to
> the row number of matrix.
> Since vectors are considered as matrices with one column, ``vector.n`` is simply
> the dimensionality of vector.

Reviewer: ```suggestion
Getting the number of matrix columns and rows will allow
you to write dimensionality-independent code. For example, this can be used to unify
2D and 3D physical simulators.

``matrix.m`` equals to the number of columns of a matrix, while ``matrix.n`` equals to
the number of rows of a matrix.
Since vectors are considered as matrices with one column, ``vector.n`` is simply
the dimensionality of the vector.
```

       x[0] = 1

* Use ``ti.static`` for forced loop unrolling:

.. code-block:: python

  @ti.kernel
  def func():
    for i in ti.static(range(4)):
        print(i)

    # is equivalent to:
    print(0)
    print(1)
    print(2)

>     print(3)

Author: I don't know, e.g.:
```rst
.. code-block::

    4tabs
```
vs
```rst
.. code-block::

  2tabs
```
Is that possible to let `ti format` take care of that too?

>     print(3)

Reviewer: > Is that possible to let `ti format` take care of that too?

We probably can, but I don't think that would be easy. We will need to write an `rst` parster that extract the code, format using `yapf`, and the put it back...

>     print(3)

Author: No, my idea is: not only parse the `code-block`, but parse the **whole `rst`**, e.g.:

```diff
- This  is a  bad line
+ This is a bad line
```
```diff
- This is a bad line that is tooooooooooo long for yapf
+ This is a bad line that is tooooooooooo
+ long for yapf
```

>     print(3)

Reviewer: You probably need a different tool. I don't think yapf can correctly handle `rst`...
